apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,362 @@
+        (uint256 tokensSold,,,,,) = projectTokenManager.getVesting(_buyer, _purchaseId);
+        projectTokenManager.revokeVesting(_buyer, _purchaseId);
+
+        // (Token manager) ~~~> project tokens ~~~> (burn 💥)
+        projectTokenManager.burn(address(projectTokenManager), tokensSold);
+
+        emit TokensRefunded(_buyer, daiToRefund, tokensSold, _purchaseId);
+    }
+
+    /**
+    * @notice Closes a sale that has reached the funding goal, sending raised funds to the fundraising pool and the beneficiary address, and initializes the Fundraising app by adding the raised funds as collateral tokens.
+    */
+    function close() public {
+        require(currentSaleState() == SaleState.GoalReached, ERROR_INVALID_STATE);
+
+        // (presale) ~~~> dai ~~~> (beneficiary)
+        uint256 daiForBeneficiary = totalDaiRaised.mul(percentFundingForBeneficiary).div(PPM);
+        require(daiToken.transfer(beneficiaryAddress, daiForBeneficiary), ERROR_DAI_TRANSFER_REVERTED);
+
+        // (presale) ~~~> dai ~~~> (pool)
+        uint256 daiForPool = daiToken.balanceOf(address(this));
+        require(daiToken.transfer(fundraisingPool, daiForPool), ERROR_DAI_TRANSFER_REVERTED);
+
+        // Initialize Fundraising app.
+        fundraisingController.addCollateralToken(
+            daiToken,
+            0, /* virtualSupply */
+            0, /* virtualBalance */
+            uint32(CONNECTOR_WEIGHT_PPM), /* reserveRatio */
+            3 * 10 ** 18, /* slippage */
+            tapRate,
+            0 /* floor */
+        );
+        fundraisingInitialized = true;
+
+        emit SaleClosed();
+    }
+
+    /*
+     * Getters
+     */
+
+    /**
+    * @notice Calculates the number of project tokens that would be obtained for `_daiAmount` dai tokens.
+    * @param _daiAmount uint256 The amount of dai tokens to be converted into project tokens.
+    */
+    function daiToProjectTokens(uint256 _daiAmount) public view returns (uint256) {
+        return _daiAmount.mul(daiToProjectTokenExchangeRate);
+    }
+
+    /**
+    * @notice Calculates the current state of the sale.
+    */
+    function currentSaleState() public view returns (SaleState) {
+        if (startDate == 0) {
+            return SaleState.Pending;
+        } else if (totalDaiRaised >= daiFundingGoal) {
+            if (fundraisingInitialized) {
+                return SaleState.Closed;
+            } else {
+                return SaleState.GoalReached;
+            }
+        } else if (_timeSinceFundingStarted() < fundingPeriod) {
+            return SaleState.Funding;
+        } else {
+            return SaleState.Refunding;
+        }
+    }
+
+    /*
+     * Internal
+     */
+
+    function _timeSinceFundingStarted() private returns (uint64) {
+        if (startDate == 0) {
+            return 0;
+        } else {
+            return getTimestamp64().sub(startDate);
+        }
+    }
+
+    function _calculateExchangeRate() private {
```

Comment:
I'm not sure to understand the rationale behind that formula ... Should the price per token be set at at fixed rate [independently of the `BancorFormula` parameters ?]

Reply:
@lkngtn would you like to comment on this?

Reply:
The rationale is that the pre-sale contract is used to bootstrap the bonding curve / market maker with initial collateral. I don't think it makes much sense for this to be tightly integrated with the market maker itself, it should be self-contained within the pre-sale logic. 

The idea is to take some parameters as inputs that are intuitive from a user perspective thinking about how much they initially need to raise, how much runway they need, whether there is an initial upfront cost that needs to be released immediate, etc in order to get a project "off the ground". 

The exchange rate offered by the pre-sale is fixed based on these parameters and intended to ensure if the pre-sale is successful that we raise a specific amount of funds and have a specific amount of tokens issued when we initialize the curve so that we can have a deterministic starting price and that we know the relative value of the collateral pool and tap rate that we start with.

Reply:
@osarrouy do you validate this?

Reply:
To have a maximum precision I would put it that way:
```
fundingGoal.mul(PPM).mul(percentSupplyOffered).div(CONNECTOR_WEIGHT_PPM).div(PPM);
```